File: frontend/src/lib/types/context.ts
export interface APODData {
  date: string;
  title: string;
  explanation: string;
  url: string;
  hdurl?: string;
  media_type: 'image' | 'video';
  copyright?: string;
}

export interface MarsPhoto {
  id: number;
  sol: number;
  earth_date: string;
  camera_name: string;
  camera_full_name: string;
  rover_name: string;
  img_src: string;
}

export interface NearEarthObject {
  id: string;
  name: string;
  nasa_jpl_url: string;
  estimated_diameter_min_m: number;
  estimated_diameter_max_m: number;
  is_potentially_hazardous: boolean;
  close_approach_date: string;
  miss_distance_km: number;
  relative_velocity_kph: number;
}

export interface HistoricalEvent {
  year: number;
  title: string;
  description: string;
  category?: string;
  wikipedia_url?: string;
  image_url?: string;
}

export interface HistoricalFigure {
  name: string;
  year: number;
  description: string;
  event_type: 'birth' | 'death';
  wikipedia_url?: string;
}

export interface HistoricalWeather {
  date: string;
  location_name?: string;
  latitude: number;
  longitude: number;
  temperature_max_c: number;
  temperature_min_c: number;
  temperature_mean_c: number;
  precipitation_mm: number;
  weather_code: number;
  weather_description: string;
  sunrise?: string;
  sunset?: string;
  daylight_hours?: number;
}

export interface TemporalContext {
  date: string;

  // Space
  apod?: APODData;
  mars_photos: MarsPhoto[];
  near_earth_objects: NearEarthObject[];

  // History
  historical_events: HistoricalEvent[];
  notable_births: HistoricalFigure[];
  notable_deaths: HistoricalFigure[];
  year_fact?: string;

  // Weather
  weather?: HistoricalWeather;
  weather_comparison: Record<number, HistoricalWeather>;

  // Insights
  cosmic_perspective?: string;
  day_summary?: string;
}

File: backend/app/schemas/context.py
from pydantic import BaseModel
from typing import Optional, List, Dict
from datetime import date


class APODResponse(BaseModel):
    date: date
    title: str
    explanation: str
    url: str
    hdurl: Optional[str] = None
    media_type: str
    copyright: Optional[str] = None


class MarsPhotoResponse(BaseModel):
    id: int
    sol: int
    earth_date: date
    camera_name: str
    camera_full_name: str
    rover_name: str
    img_src: str


class NearEarthObjectResponse(BaseModel):
    id: str
    name: str
    nasa_jpl_url: str
    estimated_diameter_min_m: float
    estimated_diameter_max_m: float
    is_potentially_hazardous: bool
    close_approach_date: date
    miss_distance_km: float
    relative_velocity_kph: float


class HistoricalEventResponse(BaseModel):
    year: int
    title: str
    description: str
    category: Optional[str] = None
    wikipedia_url: Optional[str] = None
    image_url: Optional[str] = None


class HistoricalFigureResponse(BaseModel):
    name: str
    year: int
    description: str
    event_type: str
    wikipedia_url: Optional[str] = None


class HistoricalWeatherResponse(BaseModel):
    date: date
    location_name: Optional[str] = None
    latitude: float
    longitude: float
    temperature_max_c: float
    temperature_min_c: float
    temperature_mean_c: float
    precipitation_mm: float
    weather_code: int
    weather_description: str
    sunrise: Optional[str] = None
    sunset: Optional[str] = None
    daylight_hours: Optional[float] = None


class TemporalContextResponse(BaseModel):
    date: date
    apod: Optional[APODResponse] = None
    mars_photos: List[MarsPhotoResponse] = []
    near_earth_objects: List[NearEarthObjectResponse] = []
    historical_events: List[HistoricalEventResponse] = []
    notable_births: List[HistoricalFigureResponse] = []
    notable_deaths: List[HistoricalFigureResponse] = []
    year_fact: Optional[str] = None
    weather: Optional[HistoricalWeatherResponse] = None
    weather_comparison: Dict[int, HistoricalWeatherResponse] = {}
    cosmic_perspective: Optional[str] = None
    day_summary: Optional[str] = None
